src/compiler.jl

```diff
@@ -1798,30 +1812,46 @@ end
 
 const cache = Dict{UInt, Dict{UInt, Any}}()
 
-function thunk(f::F,tt::TT=Tuple{},::Val{Split}=Val(false)) where {F<:Core.Function, TT<:Type, Split}
+function thunk(f::F,::Type{A}, tt::TT=Tuple{},::Val{Split}=Val(false)) where {F<:Core.Function, A<:Annotation, TT<:Type, Split}
     primal, adjoint = fspec(f, tt)
+
+    if A isa UnionAll
+        rt = Core.Compiler.return_type(primal.f, primal.tt)
+        rt = A{rt}
+    else
+        @assert A isa DataType
+        # Can we relax this condition?
+        @assert eltype(A) == Core.Compiler.return_type(primal.f, primal.tt)
+        rt = A
+    end
+
     # We need to use primal as the key, to lookup the right method
     # but need to mixin the hash of the adjoint to avoid cache collisions
     # This is counter-intuitive since we would expect the cache to be split
     # by the primal, but we want the generated code to be invalidated by
     # invalidations of the primal, which is managed by GPUCompiler.
-    local_cache = get!(Dict{Int, Any}, cache, hash(adjoint, UInt64(Split)))
+    local_cache = get!(Dict{Int, Any}, cache, hash(adjoint, hash(rt, UInt64(Split))))
```

Comment:
Could we get away with hash of A rather than rt?

Reply:
No this is actually more general since `A` above get's normalized to `A{rt}` meaning that a user can provide both `Active` and `Active{Float64}` and hit the same cache entry.